vinyl: fix bug in upsert squashing optimization

locker opened the pull request:
This PR fixes a bug in the Vinyl upsert optimization code that could lead to a segmentation fault. The cause of the bug is thoroughly described in https://github.com/tarantool/tarantool/issues/5080#issuecomment-1159572786.

Closes #5080

Korablev77 commented on src/box/vy_lsm.c:
```diff
@@ -888,6 +888,17 @@ vy_lsm_rotate_mem(struct vy_lsm *lsm)
 }
 
+int
+vy_lsm_check_rotate_mem(struct vy_lsm *lsm)
```
I'd call it sort of `vy_lsm_rotate_on_demand` or `vy_lsm_rotate_if_required`. Feel free to skip

locker replied:
Renamed to `vy_lsm_rotate_if_required`.

Korablev77 commented on src/box/vy_scheduler.c:
```diff
@@ -1339,8 +1339,7 @@ vy_task_dump_new(struct vy_scheduler *scheduler, struct vy_worker *worker,
 	}
 
 	/* Rotate the active tree if it needs to be dumped. */
-	if (lsm->mem->generation == scheduler->dump_generation &&
```
Now there's additional cond: `lsm->mem->space_cache_version != space_cache_version`. I believe it's probably OK, but could it lead to some undesirable side-effects? I mean here we can potentially get mem rotation on compaction..

locker replied:
Nothing bad should happen, because we won't include the `vy_mem` into dump unless its generation <= dump generation, see:

https://github.com/tarantool/tarantool/blob/47ad3bc9fc1f0e2fc8658d1e4013ce35694d2bf5/src/box/vy_scheduler.c#L1397-L1402

Moreover, the mem can be rotated by tx write anyway, because there's a time window between bumping dump generation and proceeding to dump.

Korablev77 reviewed (approved):
Why don't you squash changelog with last patch?
>NO_TEST=complicated

:)

locker commented:
> Why don't you squash changelog with last patch?

Squashed.

> > NO_TEST=complicated
> 
> :)

Added a better explanation :)

```
NO_TEST=complicated, need stress/perf test to catch bugs like this
```

locker commented:
Cherry-picked to 2.10 and 1.10.
